core/src/main/java/com/linecorp/armeria/server/PathMappings.java

```diff
@@ -118,16 +116,34 @@ public PathMappings(int cacheSize) {
         for (int i = 0; i < size; i++) {
             final Entry<PathMapping, T> e = patterns.get(i);
             final PathMapping mapping = e.getKey();
-            final PathMappingResult mappingResult = mapping.apply(path, query);
+            final PathMappingResult mappingResult = mapping.apply(mappingCtx);
             if (mappingResult.isPresent()) {
-                result = PathMapped.of(mapping, mappingResult, e.getValue());
-                break;
+                // If the result is not the best matching, we keep it and try to find the better matching again.
+                if (mappingResult.isHighestScore()) {
+                    result = PathMapped.of(mapping, mappingResult, e.getValue());
+                    break;
+                }
+
+                if (result.isPresent()) {
+                    if (mappingResult.score() > result.mappingResult().score()) {
+                        // Replace the candidate with the new one only if the score is better.
+                        // If the score is same, we respect the order of service registration.
+                        result = PathMapped.of(mapping, mappingResult, e.getValue());
+                    }
+                } else {
+                    result = PathMapped.of(mapping, mappingResult, e.getValue());
+                }
+
+                // This means that we have no more chance to get a better score. Give up and use the candidate.
+                if (mappingResult.isLowestScore()) {
+                    break;
```

Comment:
Basically our path mappings are sorted by its `complexity`, but only the annotated services have its `complexity` so they are in the front of the list. And, I added the `score` only to the `HttpHeaderPathMapping` which is used by annotated services. So, the services except for annotated services are always return the lowest score. This means we are in the stage which finds a suitable service in a registered order. So I wrote the comment like that.

If we don't do like this, we would traverse the path mapping until the end of the list even if we found the service like `THttpService`. 

I will add the comment more specifically. 😃